core/field_dropdown.js

```diff
@@ -36,698 +35,715 @@ const {Svg} = goog.require('Blockly.utils.Svg');
     }
+    const option = this.selectedOption_[0];
+    if (typeof option === 'object') {
+      return option['alt'];
+    }
+    return option;
   }
-};
 
-/**
- * Updates the dropdown arrow to match the colour/style of the block.
- * @package
- */
-FieldDropdown.prototype.applyColour = function() {
-  if (this.borderRect_) {
-    this.borderRect_.setAttribute(
-        'stroke', this.sourceBlock_.style.colourTertiary);
-    if (this.menu_) {
-      this.borderRect_.setAttribute(
-          'fill', this.sourceBlock_.style.colourTertiary);
-    } else {
-      this.borderRect_.setAttribute('fill', 'transparent');
-    }
+  /**
+   * Construct a FieldDropdown from a JSON arg object.
+   * @param {!Object} options A JSON object with options (options).
+   * @return {!FieldDropdown} The new field instance.
+   * @package
+   * @nocollapse
+   */
+  static fromJson(options) {
+    // `this` might be a subclass of FieldDropdown if that class doesn't
+    // override the static fromJson method.
+    return new this(options['options'], undefined, options);
   }
-  // Update arrow's colour.
-  if (this.sourceBlock_ && this.arrow_) {
-    if (this.sourceBlock_.isShadow()) {
-      this.arrow_.style.fill = this.sourceBlock_.style.colourSecondary;
-    } else {
-      this.arrow_.style.fill = this.sourceBlock_.style.colourPrimary;
+
+  /**
+   * Use the calculated prefix and suffix lengths to trim all of the options in
+   * the given array.
+   * @param {!Array<!Array>} options Array of option tuples:
+   *     (human-readable text or image, language-neutral name).
+   * @param {number} prefixLength The length of the common prefix.
+   * @param {number} suffixLength The length of the common suffix
+   * @return {!Array<!Array>} A new array with all of the option text trimmed.
+   */
+  static applyTrim_(options, prefixLength, suffixLength) {
+    const newOptions = [];
+    // Remove the prefix and suffix from the options.
+    for (let i = 0; i < options.length; i++) {
+      let text = options[i][0];
+      const value = options[i][1];
+      text = text.substring(prefixLength, text.length - suffixLength);
+      newOptions[i] = [text, value];
     }
+    return newOptions;
   }
-};
+}
 
 /**
- * Draws the border with the correct width.
- * @protected
+ * Dropdown image properties.
+ * @typedef {{
+ *            src:string,
+ *            alt:string,
+ *            width:number,
+ *            height:number
+ *          }}
  */
-FieldDropdown.prototype.render_ = function() {
-  // Hide both elements.
-  this.textContent_.nodeValue = '';
-  this.imageElement_.style.display = 'none';
-
-  // Show correct element.
-  const option = this.selectedOption_ && this.selectedOption_[0];
-  if (option && typeof option === 'object') {
-    this.renderSelectedImage_(
-        /** @type {!FieldDropdown.ImageProperties} */ (option));
-  } else {
-    this.renderSelectedText_();
-  }
-
-  this.positionBorderRect_();
-};
+FieldDropdown.ImageProperties;
```

Comment:
Own name as in like `exports.DropdownImageProperties`? Right now I'm kinda against exporting we don't absolutely need to export. But if it'll cause us problems with the conversion later I'm happy to add something.

Reply:
I'd probably just export it as `exports.ImageProperties`, and let the importer rename it to `DropdownImageProperties` if they had some other kind of image properties to deal with in the same module.

My preference list is, vaguely:
1. Don't export things that can be kept private.
2. If it must be exported, prefer to export it under its own name.
3. Export it as a property of another exported value only if necessary for backwards compatibility or if it is somehow very tightly coupled to that other exported value.

Is `ImageProperties` used outside this module?

Reply:
> Is ImageProperties used outside this module?

It is not. It is only used to provide a type to the FieldDropdown's [renderSelectedImage_](https://github.com/google/blockly/blob/c9be8d984d6cc459917f2a777e03ba858f2e623d/core/field_dropdown.js#L608) method. Which turns out to actually be private.

So in that case I believe it can be removed from the public-facing API, and just declared as a `const` within that file? Does that sound right to you?

Reply:
> So in that case I believe it can be removed from the public-facing API, and just declared as a `const` within that file? Does that sound right to you?

Yes.